actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,91 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      unless partial_cache.empty?
+        ActionView::Base.logger.try :debug, "#{partial_cache.size} partials " \
+          "multi read from cache."
+      end
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }.values
+      rendered_partials = @collection.any? ? yield.dup : []
+
+      unless rendered_partials.empty?
+        ActionView::Base.logger.try :debug, "#{rendered_partials.size} partials" \
+          " will be written to cache."
+      end
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      ActionController::Base.perform_caching && @options.fetch(:cache, automatic_cache_eligible?)
+    end
+
+    def automatic_cache_eligible?
+      single_template_render? && !callable_cache_key? && template_eligible?
+    end
+
+    def single_template_render?
+      @template # Template is only set when a collection renders one template.
+    end
+
+    def callable_cache_key?
+      @options[:cache].respond_to?(:call)
+    end
+
+    def template_eligible?
+      @template.eligible_for_collection_caching?(as: @options[:as]).tap do
+        ActionView::Base.logger.try :debug,
+          "Collection automatically caches #{@template.virtual_path}, so next render uses a multi read for any templates rendered now."
+      end
+    end
+
+    def collection_by_cache_keys
+      seed = callable_cache_key? ? @options[:cache] : ->(i) { i }
+
+      @collection.each_with_object({}) do |item, hash|
+        hash[expanded_cache_key(seed.call(item))] = item
+      end
+    end
+
+    def expanded_cache_key(key)
+      key = caching_controller.fragment_cache_key(@view.cache_fragment_name(key))
+      key.frozen? ? key.dup : key # #read_multi & #write may require mutability, Dalli 2.6.0.
+    end
+
+    def caching_controller
+      @_caching_controller ||= defined?(@_controller) ? @_controller : ActionController::Base.new
```

Comment:
:+1:, we're going to have to extract `fragment_cache_key` into a separate module included in controllers and views.

Then this would just be `@view.fragment_cache_key`